File: app/js/engine/tile.ts
/// <reference path="../../public/third-party-js/lodash.d.ts" />
/// <reference path="building-type.ts" />
/// <reference path="terrain-type.ts" />

module Miner {
  export class Tile {
    constructor(
      public x: number,
      public y: number,
      public terrainType: TerrainType,
      public buildingType: BuildingType,
      public remainingBuildingConstructionDays: number) {
    }

    distanceToTile(tile: Tile) {
      return this.distanceTo(tile.x, tile.y);
    }

    distanceTo(x: number, y: number) {
      return Tile.distanceBetween(x, y, this.x, this.y);
    }

    isAdjacentTo(tile: Tile) {
      return this.distanceToTile(tile) === 1;
    }

    isConstructedBuilding() {
      return this.buildingType !== null && this.remainingBuildingConstructionDays === 0;
    }

    isUnderConstruction() {
      return this.buildingType !== null && this.remainingBuildingConstructionDays > 0;
    }
  }

  export module Tile {
    export function distanceBetween(x1: number, y1: number, x2: number, y2: number) {
      var z1 = 0 - x1 - y1;
      var z2 = 0 - x2 - y2;
      return _.max([
        Math.abs(x1 - x2),
        Math.abs(y1 - y2),
        Math.abs(z1 - z2)
      ]);
    }

    export function fromJSON(t: {}) {
      var tile = new Tile(null, null, null, null, null);
      _.merge(tile, t);
      return tile;
    }
  }
}

File: libgdx2/miner/core/src/org/quach/miner/engine/Tile.java
package org.quach.miner.engine;

public class Tile {
  public final int x;
  public final int y;
  public final TerrainType terrainType;

  private BuildingType buildingType;
  private int remainingBuildingConstructionDays;

  public Tile(
    final int x,
    final int y,
    final TerrainType terrainType,
    final BuildingType buildingType,
    final int remainingBuildingConstructionDays) {
    this.x = x;
    this.y = y;
    this.terrainType = terrainType;
    this.buildingType = buildingType;
    this.remainingBuildingConstructionDays = remainingBuildingConstructionDays;
  }

  public int distanceToTile(final Tile tile) {
    return distanceTo(tile.x, tile.y);
  }

  public int distanceTo(int x, int y) {
    return distanceBetween(x, y, x, y);
  }

  public boolean isAdjacentTo(final Tile tile) {
    return distanceToTile(tile) == 1;
  }

  public boolean isConstructedBuilding() {
    return buildingType != null && remainingBuildingConstructionDays == 0;
  }

  public boolean isUnderConstruction() {
    return buildingType != null && remainingBuildingConstructionDays > 0;
  }

  public static int distanceBetween(final int x1, final int y1, final int x2, final int y2) {
    final int z1 = 0 - x1 - y1;
    final int z2 = 0 - x2 - y2;
    return Math.max(Math.max(Math.abs(x1 - x2), Math.abs(y1 - y2)), Math.abs(z1 - z2));
  }
}
